nova/metrics.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2023 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from enum import auto
+from enum import Enum
+import logging
+from typing import Union
+
+import statsd
+
+import nova.conf
+
+LOG = logging.getLogger(__name__)
+CONF = nova.conf.CONF
+
+
+class MetricType(Enum):
+    """https://statsd.readthedocs.io/en/v4.0.1/types.html"""
+    COUNTER = auto()
+    TIMER = auto()
+    GAUGE = auto()
+    SET = auto()
+
+
+_METRICS_PREFIX = "openstack.compute"
+
+
+def _send_metric(tag: str, value: Union[int, float, None],
+                 m_type: MetricType = MetricType.GAUGE):
+    client = statsd.StatsClient(CONF.statsd_host, CONF.statsd_port)
```

Comment:
I know, it is rather rare that the metric is send right now, so the changes are more stylistic, than me expecting real problems now.

I would suggest to change the code in two ways. 
1. Create a module variable for the statsd client, and allow re-use over function calls to avoid the overhead of resolving and creating a socket for each metric call.
2.  [Close the client/socket](https://github.com/jsocol/pystatsd/blob/c687a4e6bcad1314088cf4934580a8bf46290a8a/statsd/client/udp.py#L47) after using it. Preferably with [contextlib.closing](https://docs.python.org/3/library/contextlib.html#contextlib.closing)

Reply:
My preferred  variant would be (1) anyway. If we have to bump the statsd client, even ore so.